File: 872. Leaf-Similar Trees/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
  def leafSimilar(self, root1, root2):
    """
    :type root1: TreeNode
    :type root2: TreeNode
    :rtype: bool
    """
    def dfs(current, array):
      if current.left == None and current.right == None:
        array.append(current.val)
      if current.left != None:
        dfs(current.left, array)
      if current.right != None:
        dfs(current.right, array)
    root1Leaf = []
    root2Leaf = []
    dfs(root1, root1Leaf)
    dfs(root2, root2Leaf)
    return root1Leaf == root2Leaf

File: 872. Leaf-Similar Trees/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  void dfs(TreeNode* current, vector<int> &array){
    if (current->left == NULL && current->right == NULL){
      array.push_back(current->val);
    }
    if (current->left != NULL){
      dfs(current->left, array);
    }
    if (current->right != NULL){
      dfs(current->right, array);
    }
  }
public:
  bool leafSimilar(TreeNode* root1, TreeNode* root2) {
    vector<int> root1Leaf;
    vector<int> root2Leaf;
    dfs(root1, root1Leaf);
    dfs(root2, root2Leaf);
    return root1Leaf == root2Leaf;
  }
};